### PR TITLE
programming on data.table

### PR DESCRIPTION
**Pull request by jangorecki**
This PR is meant to introduce new (base-R like) interface for parametrizing data.table queries. #2655
It uses base R `substitute` style substitution, but extends it for substitution of names of the argument in a function calls. This allows to give names of a columns from a variables as well, which is not possible using just base R substitute.
```r
# data
DT = data.table(x = 1:5, y = 5:1)

# parameters
in_col_name = "x"
fun = "sum"
fun_arg1 = "na.rm"
fun_arg1val = TRUE
out_col_name = "sum_x"

# query
DT[, .(out_col_name = fun(in_col_name, fun_arg1=fun_arg1val))]

# desired query
DT[, .(sum_x = sum(x, na.rm=TRUE))]

# proposed API
DT[, .(out_col_name = fun(in_col_name, fun_arg1=fun_arg1val)),
  env = list(in_col_name=in_col_name, fun=fun, fun_arg1=fun_arg1, fun_arg1val=fun_arg1val, out_col_name=out_col_name)]
```

working example, checkout to `programming` branch, start `R`
```r
cc(F) # build and attach pkg

substitute2(
  .(out_col_name = fun(in_col_name, fun_arg1=fun_arg1val)),
  env = list(
    in_col_name = "x",
    fun = "sum",
    fun_arg1 = "na.rm",
    fun_arg1val = TRUE,
    out_col_name = "sum_x"
  )
)
#.(sum_x = sum(x, na.rm = TRUE))
```
Feedback welcome!

### CODE REVIEW

**Comment by renkun-ken** [comment]
Really like the solution here. Desperately need programming with data.table in my code. The design looks concise and powerful enough for most of my use cases.

File: R/programming.R [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  env.names = names(env)
+  if (is.null(env.names)) {
+    stop("'env' argument does not have names")
+  } else if (!all(nzchar(env.names))) {
+    stop("'env' argument has an zero char names")
+  }
+  if (isTRUE(char.as.name)) {
+    char = vapply(env, is.character, FALSE)
+    if (any(char)) {
+      if (any(non.scalar.char <- lengths(env[char])!=1L)) {
```
Comment (MichaelChirico) [inline R/programming.R]: `lengths` is not 3.1.0 right? Maybe we should just backport it like we did `trimws`

**Comment by renkun-ken** [comment]
I'm wondering if this `substitute2` will be applied on the whole expression or only `j`, or `i`, `j`, `by`, `keyby`?

Some of my use case is like

```r
dt[col > mean(col), lapply(.SD, avg, na.rm = TRUE),
  keyby = .(symbol, date, group), .SDcols = paste0("x", 1:10),
  env = list(col = "col1", avg = "median", group = "UNIVERSE")]
```

which needs the substitution be done in `i`, `j`, `by`/`keyby`.

File: R/programming.R (hunk 1 above)
Comment (jangorecki) [reply R/programming.R]: Not big issue about vapply instead, but what I would like to have more is to coerce 'env' to environment and use that rather than a list. This is how base R uses 'env' argument in 'substitute'. Anyway need to measure performance before deciding which one to use.

**Comment by jangorecki** [comment]
It could be easily applied to any argument that we now use substitute, and yes i, j, by, keyby all make sense. This also includes parameterizing a data.table to join to in 'i' argument, or special symbols like .SD.
Another functionality that user could actually wants is a splice feature of bquote, but I was not looking into that yet. For now not planning to put this in PR as API for that is not easy to figure out.

**Comment by MichaelChirico** [comment]
Not a blocker, but just to make note:

We'll probably want a vignette with a nice worked example/practical use case motivating/demonstrating the goals of this before release

**Comment by jangorecki** [comment]
Having good `substitute2` function it was quite easy to plug it into `[.data.table`. It is now implemented for `i`, `j` and `by` (or `keyby`). I will work on the manual, not sure about vignette yet. I would appriciate contribution of use cases, so we can see how much it can cover, and if it covers as we expect.

**Comment by renkun-ken** [comment]
Thanks for the great work! Use cases, Of course! I’ll provide as many use cases as I have soon.

**Comment by renkun-ken** [comment]
Here are my use cases that will be greatly simplified by this:

```r
library(data.table)

dt <- data.table(symbol = rep(1:3, each = 10))
dt[, date := seq_len(.N), by = symbol]
dt[, grp1 := sample.int(2, .N, replace = TRUE)]
dt[, grp2 := sample.int(3, .N, replace = TRUE)]
dt[, x0 := rnorm(.N)]
dt[, x1 := rnorm(.N)]
dt[, x2 := rnorm(.N)]
dt[, x3 := rnorm(.N)]
dt[, y1 := rnorm(.N)]
dt[, y2 := rnorm(.N)]
dt[, y3 := rnorm(.N)]

# Use case: apply same formula to different set of columns

xs <- c("x", "y")
for (x in xs) {
  dt[, RATIO := (R3 - R2) * (R2 - R1) * (R3 - R1) / sqrt(R1^2 + R2^2 + R3^2),
    env = list(RATIO = paste0(x, "_ratio"),
      R1 = paste0(x, 1), R2 = paste0(x, 2), R3 = paste0(x, 3)),
    verbose = TRUE]
}

# Use case: daily correlation of user input features

daily_cor <- function(data, x, y) {
  data[, .(cor = cor(x, y)),
    keyby = date,
    env = list(x = x, y = y),
    verbose = TRUE]
}

daily_cor(dt, "x0", "y2")

# Use case: group cor comparison of user input features

group_cor <- function(data, x, y, g) {
  cor_dt <- data[, lapply(.SD, function(x) cor(x, Y)),
    keyby = .(group = GROUP),
    .SDcols = x,
    env = list(Y = y, GROUP = g),
    verbose = TRUE]
  melt.data.table(cor_dt, id.vars = "group", measure.vars = x,
    variable.name = "x", value.name = "cor", variable.factor = FALSE)
}

dt1 <- group_cor(dt, c("x0", "x1", "x2"), "y1", "grp1")
dt2 <- group_cor(dt, c("x0", "x1", "x2"), "y1", "grp2")

## post aggregation with known colnames

dt1[, .(min = min(cor), mean = mean(cor), max = max(cor)), keyby = x]
dt2[, .(min = min(cor), mean = mean(cor), max = max(cor)), keyby = x]


yn <- c(1, 5, 10, 20)
ycols <- paste0("y", yn)

ydt <- data.table(symbol = rep(1:3, each = 100))
ydt[, date := seq_len(.N), by = symbol]
ydt[, ret := rnorm(.N)]
ydt[, (ycols) := shift(ret, yn, type = "lead"), by = symbol]

xdt <- data.table(symbol = rep(1:2, each = 20))
xdt[, date := seq_len(.N), by = symbol]
xdt[, x1 := rnorm(.N)]
xdt[, x2 := rnorm(.N)]

# Use case: cor between each x and a single y
cor_xy <- function(xdt, ydt, x, y) {
  xdt[ydt, y := Y, on = .(symbol, date),
    env = list(Y = y),
    verbose = TRUE
  ]
  on.exit(xdt[, y := NULL])
  cor_dt <- xdt[, lapply(.SD, cor, y = y), keyby = symbol, .SDcols = x]
  cor_dt
}

cor_xy(xdt, ydt, c("x1", "x2"), "y10")

# Use case: cor between each pair of x and y
cor_xy2 <- function(xdt, ydt, x, y) {
  res <- rbindlist(lapply(y, function(yi) {
    xdt[ydt, y := Y, on = .(symbol, date),
      env = list(Y = yi)
    ]
    on.exit(xdt[, y := NULL])
    rbindlist(lapply(x, function(xi) {
      xdt[, .(x = xi, y = yi, cor = cor(X, y)), keyby = symbol,
        env = list(X = xi)]
    }))
  }))
}

cor_dt <- cor_xy2(xdt, ydt, c("x1", "x2"), ycols)

# Use case: cor matrix of existing columns and dynamically in-place merged columns
cor_xy3 <- function(xdt, ydt, x, y) {
  # still can't easily program without constructing a call
  xdt[ydt, `:=`(
    y1 = y1,
    y5 = y5,
    y10 = y10,
    y20 = y20
  ), on = .(symbol, date)]
  on.exit(xdt[, (ycols) := NULL])
  xdt[, cor(.SD), .SDcols = c(x, y)]
}

cor_xy3(xdt, ydt, c("x1", "x2"), ycols)
```

**Comment by jangorecki** [comment]
I am thinking about removing 2 arguments from substitute2 API, I would appriciate feedback, if anyone would found them useful:

- `sub.names` - now when setting to false, it will not substitute names, so will make it behave like base R `substitute`, the difference to base substitute will be that character elements in `env` will still be preprocessed to symbols, unless...
- `char.as.name` - parameter to control if we want to preprocess character elements in `env` to symbols, by default TRUE, unless `AsIs` class is passed to `env`. We are able to handle behaviour of that by using `env = I(list(...))` or `env = list(v = I("char"))`, so argument seems reduntant.

Using both `char.as.name=FALSE` and `sub.names=FALSE` will make it behaves like base R substitute. So, do we need those options?

**Comment by jangorecki** [comment]
@renkun-ken I pushed your use cases in 507e253. I used `dput` of the current answers to produce expected values, could you please verify you are getting same numbers in your workflow?

----

Additionally, extra parameters mentioned in previous comment (`char.as.name` and `sub.names`) has been removed.

**Comment by renkun-ken** [comment]
@jangorecki Thanks for adding my use cases.

Please notice that the `cor_xy3` does not work with this programming feature yet as I comment in the code

```r
# still can't easily program without constructing a call
  xdt[ydt, `:=`(
    y1 = y1,
    y5 = y5,
    y10 = y10,
    y20 = y20
  ), on = .(symbol, date)]
```

I was hoping for something that allows me to also dynamically generate `:=` args using input `y` rather than hard code like this, as I requested in https://github.com/Rdatatable/data.table/issues/3184.

**Comment by renkun-ken** [comment]
@jangorecki To confirm the results are correct, it seems I should provide my original solution to these use cases where this programming feature is not used.

I'll update the code soon.

**Comment by renkun-ken** [comment]
@jangorecki My use cases are updated to include the old solutions (mostly use `.SD[[1]]` to refer to the target column), and all results are consistent. The `cor_xy3` use case is removed.

```r
library(data.table)

dt <- data.table(symbol = rep(1:3, each = 10))
dt[, date := seq_len(.N), by = symbol]
dt[, grp1 := sample.int(2, .N, replace = TRUE)]
dt[, grp2 := sample.int(3, .N, replace = TRUE)]
dt[, x0 := rnorm(.N)]
dt[, x1 := rnorm(.N)]
dt[, x2 := rnorm(.N)]
dt[, x3 := rnorm(.N)]
dt[, y1 := rnorm(.N)]
dt[, y2 := rnorm(.N)]
dt[, y3 := rnorm(.N)]

# Use case: apply same formula to different set of columns

## old

xs <- c("x", "y")
for (x in xs) {
  dt[, paste0(x, "_ratio") := (.SD[[3]] - .SD[[2]]) * (.SD[[2]] - .SD[[1]]) * (.SD[[3]] - .SD[[1]]) / sqrt(.SD[[1]]^2 + .SD[[2]]^2 + .SD[[3]]^2),
    .SDcols = paste0("x", 1:3)]
}

## new

xs <- c("x", "y")
for (x in xs) {
  dt[, RATIO := (R3 - R2) * (R2 - R1) * (R3 - R1) / sqrt(R1^2 + R2^2 + R3^2),
    env = list(RATIO = paste0(x, "_ratio"),
      R1 = paste0(x, 1), R2 = paste0(x, 2), R3 = paste0(x, 3)),
    verbose = TRUE]
}

# Use case: daily correlation of user input features

## old

daily_cor <- function(data, x, y) {
  data[, .(cor = cor(.SD[[1]], .SD[[2]])),
    keyby = date,
    .SDcols = c(x, y)
  ]
}

## new

daily_cor <- function(data, x, y) {
  data[, .(cor = cor(x, y)),
    keyby = date,
    env = list(x = x, y = y),
    verbose = TRUE]
}

daily_cor(dt, "x0", "y2")

# Use case: group cor comparison of user input features

## old

group_cor <- function(data, x, y, g) {
  data[, .y := .SD[[1]], .SDcols = y]
  data[, .group := .SD[[1]], .SDcols = g]
  on.exit(data[, c(".y", ".group") := NULL])
  cor_dt <- data[, lapply(.SD, function(x) cor(x, .y)),
    keyby = .(group = .group),
    .SDcols = x]
  melt.data.table(cor_dt, id.vars = "group", measure.vars = x,
    variable.name = "x", value.name = "cor", variable.factor = FALSE)
}

## new

group_cor <- function(data, x, y, g) {
  cor_dt <- data[, lapply(.SD, function(x) cor(x, Y)),
    keyby = .(group = GROUP),
    .SDcols = x,
    env = list(Y = y, GROUP = g),
    verbose = TRUE]
  melt.data.table(cor_dt, id.vars = "group", measure.vars = x,
    variable.name = "x", value.name = "cor", variable.factor = FALSE)
}

dt1 <- group_cor(dt, c("x0", "x1", "x2"), "y1", "grp1")
dt2 <- group_cor(dt, c("x0", "x1", "x2"), "y1", "grp2")

## post aggregation with known colnames

dt1[, .(min = min(cor), mean = mean(cor), max = max(cor)), keyby = x]
dt2[, .(min = min(cor), mean = mean(cor), max = max(cor)), keyby = x]


yn <- c(1, 5, 10, 20)
ycols <- paste0("y", yn)

ydt <- data.table(symbol = rep(1:3, each = 100))
ydt[, date := seq_len(.N), by = symbol]
ydt[, ret := rnorm(.N)]
ydt[, (ycols) := shift(ret, yn, type = "lead"), by = symbol]

xdt <- data.table(symbol = rep(1:2, each = 20))
xdt[, date := seq_len(.N), by = symbol]
xdt[, x1 := rnorm(.N)]
xdt[, x2 := rnorm(.N)]

# Use case: cor between each x and a single y

## old

cor_xy <- function(xdt, ydt, x, y) {
  ydt[, .y := .SD[[1]], .SDcols = y]
  on.exit(ydt[, .y := NULL])
  xdt[ydt, y := .y, on = .(symbol, date)]
  on.exit(xdt[, y := NULL], add = TRUE)
  cor_dt <- xdt[, lapply(.SD, cor, y = y), keyby = symbol, .SDcols = x]
  cor_dt
}

## new

cor_xy <- function(xdt, ydt, x, y) {
  xdt[ydt, y := Y, on = .(symbol, date),
    env = list(Y = y),
    verbose = TRUE
  ]
  on.exit(xdt[, y := NULL])
  cor_dt <- xdt[, lapply(.SD, cor, y = y), keyby = symbol, .SDcols = x]
  cor_dt
}

cor_xy(xdt, ydt, c("x1", "x2"), "y10")

# Use case: cor between each pair of x and y

## old

cor_xy2 <- function(xdt, ydt, x, y) {
  res <- rbindlist(lapply(y, function(yi) {
    ydt[, .y := .SD[[1]], .SDcols = yi]
    on.exit(ydt[, .y := NULL])
    xdt[ydt, y := .y, on = .(symbol, date)]
    on.exit(xdt[, y := NULL], add = TRUE)
    rbindlist(lapply(x, function(xi) {
      xdt[, .(x = xi, y = yi, cor = cor(.SD[[1]], y)), keyby = symbol,
        .SDcols = xi]
    }))
  }))
}

## new

cor_xy2 <- function(xdt, ydt, x, y) {
  res <- rbindlist(lapply(y, function(yi) {
    xdt[ydt, y := Y, on = .(symbol, date),
      env = list(Y = yi)
    ]
    on.exit(xdt[, y := NULL])
    rbindlist(lapply(x, function(xi) {
      xdt[, .(x = xi, y = yi, cor = cor(X, y)), keyby = symbol,
        env = list(X = xi)]
    }))
  }))
}

cor_dt <- cor_xy2(xdt, ydt, c("x1", "x2"), ycols)

# Use case: fill abnormal values of multiple columns
nadt <- data.table(x1 = c(1, 2, NA, Inf), x2 = c(2, NA, 3, Inf), x3 = c(NA, 1, 2, 0))

## old

dt_fill <- function(data, columns, selector, fill) {
  selector <- match.fun(selector)
  for (col in columns) {
    sel <- selector(data[[col]])
    data[sel, (col) := fill]
  }
}

## new

dt_fill <- function(data, columns, selector, fill) {
  selector <- match.fun(selector)
  for (col in columns) {
    data[selector(X), X := fill,
      env = list(X = col)]
  }
}

dt_fill(nadt, c("x1", "x2", "x3"), is.na, 0)
dt_fill(nadt, c("x1", "x2", "x3"), is.infinite, 0)
```

**Comment by jangorecki** [comment]
@renkun-ken thanks, but I think the current commit is fine, we test there all what needs to be tested, isn't it?
also re `cor_xy3`, why you removed that? I improved the test instead.

**Comment by renkun-ken** [comment]
@jangorecki Thanks for updating `cor_xy3` with constructed call. I should use `y` instead of `ycols` inside the function to show that user could input some `y` instead of all `ycols`.

```r
cor_xy3 = function(xdt, ydt, x, y) { ## cor matrix of existing columns and dynamically in-place merged columns
  qc = as.call(lapply(setNames(c(":=", y), c("", y)), as.name))
  xdt[ydt, j, on = .(symbol, date),
      env = list(j=qc)]
  on.exit(xdt[, (y) := NULL])
  xdt[, cor(.SD), .SDcols = c(x, y)]
}
```

Everything else looks good to me. I'm completely fine with removing my last long comment if you think the test case should be hard coded numbers instead of comparing the output of the old and new way.

**Comment by tdeenes** [comment]
Maybe this use case has not been covered yet:

```r
# join programmatically
dt1 <- data.table(x = letters[1:5], y = 1:5)
dt2 <- data.table(x = letters[1:3], y = 11:13)
target_v <- "y"
source_v <- paste0("i.", target_v)
on_v <- "x"
dt1[
  dt2, 
  target_v := source_v, 
  on = on_v, 
  env = list(target_v = target_v, source_v = source_v)
]
```

However, I do not know whether we can have multiple target/source variables with `:=` without constructing a call.

**Comment by jangorecki** [comment]
@tdeenes it looks it is covered already, note that `on` argument is handled as before, and there are no plans to substitute inside of it.
```r
dt1 <- data.table(x = letters[1:5], y = 1:5)
dt2 <- data.table(x = letters[1:3], y = 11:13)
target_v <- "y"
source_v <- paste0("i.", target_v)
on_v <- "x"
dt1[
  dt2, 
  target_v := source_v, 
  on = on_v, 
  env = list(target_v = target_v, source_v = source_v),
  verbose = TRUE
]
#Argument 'j'  after substitute: `:=`(y, i.y)
#Argument 'i'  after substitute: dt2
#i.x has same type (character) as x.x. No coercion needed.
#forder.c received 5 rows and 2 columns
#Calculated ad hoc index in 0.000s elapsed (0.001s cpu) 
#Starting bmerge ...
#forder.c received 3 rows and 2 columns
#bmerge done in 0.036s elapsed (0.035s cpu) 
#Constructing irows for '!byjoin || nqbyjoin' ... 0.000s elapsed (0.000s cpu) 
#Detected that j uses these columns: y,i.y 
#Assigning to 3 row subset of 5 rows
#RHS_list_of_columns == false
dt1
#        x     y
#1:      a    11
#2:      b    12
#3:      c    13
#4:      d     4
#5:      e     5
```

**Comment by tdeenes** [comment]
@jangorecki  Sorry for the ambiguous wording: I wanted to add this example as a further potential use case besides those that @renkun-ken has already added. Dealing with 'i.<some variable in DT in i>' programmatically was not easy previously, and can be solved in the new API in a straightforward manner.

**Comment by jangorecki** [comment]
Interface provided in this PR addresses multiple issues, many of them asking for get/mget/eval interfaces: #1985, #2589, #2884, #2816, #750, #2432, #628, #4247.

I quoted @mmaechler in new vignette, please let me know if you would not be OK about being cited there.

**Comment by jangorecki** [comment]
This PR seems to be failing our strict check due to extra NOTE
```
 * checking installed package size ... NOTE
   installed size is  5.0Mb
   sub-directories of 1Mb or more:
     tests   1.8Mb
```

File: R/programming.R [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  as.call(c(quote(list), list2lang(x)))
+}
+
+substitute2 = function(expr, env) {
+  if (missing(env)) {
+    stop("'env' must not be missing")
+  } else if (is.null(env)) {
+    # null is fine, will be escaped few lines below
+  } else if (is.environment(env)) {
+    env = as.list(env, all.names=TRUE, sorted=TRUE)
```
[the remark this answers is not in the document]
Comment (jangorecki) [reply R/programming.R:50]: Interesting question...

TL;DR
works as documented.

----

Following chunk of code presents current difference vs base R substitute.
```r
e = new.env()
f = function() {message("f"); "V1"}
makeActiveBinding("x", f, e)
e$x
#f
#[1] "V1"
substitute2(list(x), e)
#list(function() {message("f"); "V1"})
substitute(list(x), e)
#f
#list("V1")
```
To have best consistency to base R substitute, we should convert list to an environment rather than environment to a list.
Another benefit of that would be to avoid repeating C `chmatch` (one time for each nested, named list) that we now use to compare names from `expr` with names from `env`. When having environment object we can use `findVarInFrame3`. Although now sure which one will be faster.
On the other hand having environment object would be more complex to pre-process (character to symbol, list to list call, recursively), as we would need to take care of the promises to not evaluate them (see test 2.820).
In the end, we somehow document it by: `env` is _List, or an environment that will be coerced to list, from which variables will be taken to inject into `expr`_.
Considering the use case is rather uncommon, I wouldn't much bother now about addressing it. Did you actually use active bindings that you would like to use passing in an enviroment for substitution?

[the remark this answers is not in the document]
Comment (renkun-ken) [reply R/programming.R:50]: Thanks for your detailed reply! Just curious about the consistency, personally, I don't have any use case of active bindings.

I think the documentation looks good to make it clear.

**Comment by renkun-ken** [comment]
Another question, should `env` also works with `on=` since it also supports symbols and expressions?

**Comment by jangorecki** [comment]
IMO doesn't have to. `on` has full flexibility already, by using its character interface rather than NSE. Moreover interface for joining will be eventually improved. @arunsrinivasan had a nice proposal for new API but I think he haven't yet put it on github. AFAIR it would encapsulate all join related arguments (on, mult, roll, etc.) into single argument.

**Comment by renkun-ken** [comment]
Maybe I'm missing something? I'm wondering how I could make the `on=` use case easily work without `env` support since `on = .(symbol, grp = g)` has to substitute both `grp` and `g`. The `g` case can be easily resolved using character interface but I'm not sure about the `grp` case which seems that only the `env` interface using `substitute2` could replace it easily.

Also, if `env` works for `i`, `j`, `by`, and `keyby` which all support NSE, user seems to be very likely to assume that it also works with `on` which also supports NSE, for consistency.

Looking forward to @arunsrinivasan's proposal.

**Comment by jangorecki** [comment]
@renkun-ken by character interface to `on` I mean this one
```r
dt[dt1, z1 := z, on = c("symbol", paste("grp1", "g2", sep="=="))]
dt[dt1, z2 := z, on = c("symbol", paste("grp2", "g1", sep="=="))]
```

**Comment by renkun-ken** [comment]
@jangorecki Thanks for your reminder on the character interface of `on`, never really use it before. Just feel a little inconsistent here while not all NSE arguments support `env`, but I'm totally okay with it.

**Comment by jangorecki** [comment]
@mattdowle dot dot prefix is not yet mentioned in new vignette. I skip it because it is not yet fully substituted from (nested) calls. I think it is best to put it to vignette when it will be well supported, otherwise users will just expect `DT[..var == 1L]` or `DT[, sum(..var)]` to work.

**Comment by brodieG** [comment]
Looks like great new functionality.  I haven't had a chance to fully reflect on it, but one thing I wonder is if "env" is the right term to use for the parameter.  I realize `substitute` uses it, but the semantics are different enough (though intent similar) that it may cause more confusion.   Substitute gets away with it due to the special case treatment of promises.  Something like `rename` seems more appropriate, though the possibility of `I(obj)` complicates that a bit (maybe `replace`, or too wordy).  I guess even `sub` for `substitute`.

You might also be interested in [WinVector's macro discussions](https://github.com/WinVector/wrapr/blob/master/extras/MacrosInR.md) just in case you haven't seem them.

While I'm providing unasked-for-feedback, I might as well note that this seems like a good opportunity to ensure there is consistent programmable NSE across all the interfaces in data.table, as @renkun-ken suggests.  Just because a different programmable interface exists doesn't take away from the value of a consistent interface across all.

Anyhow, feel free to ignore all this.

**Comment by MichaelChirico** [comment]
@brodieG anyone's, but especially your feedback is always welcome, never hesitate!

**Comment by jangorecki** [comment]
@brodieG Thank you for your feedback.

> I wonder is if "env" is the right term to use for the parameter.

I don't think we can figure out a single best arg name here. My preference is to invent less and re-use what is already in base R, as long as it fits. The use case is pretty much the same, so IMO it fits here quite well.

> You might also be interested in WinVector's macro discussions just in case you haven't seem them.

Haven't seen it before, thanks for pointing out. It is actually an amazing work made by @JohnMount there. I would love to have a feedback on this PR from him as well!

> consistent programmable NSE across all the interfaces in data.table, as @renkun-ken suggests. 

IMO the less NSE the better. We still want to have NSE, so it is easy to subset in `i`, do any computation easily in `j`, group by on-the-fly computed columns in `by`. My preference is to avoid NSE where it doesn't really solve new problem (saving extra quotes `"` in a character scalar vs symbol is not a problem really). So as long as we don't have on-the-fly computated columns in join conditions, there is not much to gain in having NSE there. We have it there already, yes, we have it in `.SDcols` as well, just to handle `:` calls like `id1:id3`. Each single NSE interface brings extra challenges and extra maintenance effort. Doing substitution in `.SDcols` sounds definitely like a bad idea, because the goal of `.SDcols` is already to substitute `.SD` in `j`. There are multiple pending issues related to extending `join` interface, so the `on` argument, while it will stay in `[.data.table` probably forever, will be eventually retired. On the other hand `i`, `j`, `by` are unlikely to be retired ever. Note that nothing stops a user from having substitution more globally, on every single argument in data.table, just by wrapping it from the outside.
```r
d = as.data.table(iris)
substitute2(
  d[a = x, b = y, c = z],
  env = list(
    a = "i", x = "d2",
    b = "j", y = as.list(names(d)),
    c = "on", z = I("Species")
  )
)
#d[i = d2, j = list(Sepal.Length, Sepal.Width, Petal.Length, Petal.Width, Species), on = "Species"]
```

**Comment by MichaelChirico** [comment]
@jangorecki random thought:

It seems like `env` as implemented here can fully subsume all the use cases for `with=FALSE`, is that right? If so, should we explicitly include with this PR an intent to make `with=FALSE` obsolete / deprecated in favor of `env`?

See e.g. #3916 for a bunch of `with=FALSE` use cases that couldn't be eliminated by `..var` syntax but might be handled by `env`

**Comment by jangorecki** [comment]
@MichaelChirico personally I liked `with=FALSE` as it is straightforward way to make `j` behave like in `[.data.frame`. It is also very common use case, while `env` is targetting more difficult cases, thus I wouldn't advocate for replacing `with=FALSE` by `env`.

**Comment by renkun-ken** [comment]
I agree with @jangorecki. My use cases of `with = FALSE` is that I need simple and exact `[.data.frame` behavior to use a calculated character vector to select columns, and `with = FALSE` is a very simple and clear sign of such intention. If `env` is used instead in this case, the intention is no longer straightforward so that I have to take a closer look at the code to figure out what's the code is trying to do since `env` does much more than `with = FALSE`.

**Comment by MichaelChirico** [comment]
@renkun-ken have a look at #3916 -- `..col` & the recent literal-`character` / literal-`numeric` column selection combined cover about 80-90% of the use cases I've seen.

The rest are already somewhat involved, which is why I thought to use `env` instead of `with=F`. Interested to hear your use cases that are in the gap between `..col` and `env` where `with=FALSE` would be essential

**Comment by MichaelChirico** [comment]
Here would be a good place for `substitute2`:

https://github.com/Rdatatable/data.table/blob/b417000b570133d1d014a4745163142d0c68e195/R/as.data.table.R#L215

To solve #4468 concisely

**Comment by jan-glx** [comment]
Why not use the quasi quotation syntax of tidyeval?

**Comment by jangorecki** [comment]
@jan-glx I don't think there is a need for an extra dependency for something that base R can handle well. The only more tricky thing in base R metaprogramming is to substitute names of arguments, but this not less tricky with `rlang` where you need to use rlang's `:=` operator, that would conflict with DT use of `:=` . Are there any advantages of using rlang over base R here? This PR brings base R way into `[`.
Another reason is API stability, I don't see that to be priority in tidyverse packages and personally I would prefer to use base R just because of that. I don't want to criticize here. There are always two sides, tidyverse goes towards "always improving", while base R or data.table prefers to be more conservative and provide longer lifecycle for their code.

**Comment by jan-glx** [comment]
I feel like there would not be a conflict, just a - perhaps confusing - dual use of `:=` :
This would be achieved by a slightly different perspective on the problem: Instead of providing meta-programming for data.table queries, one would just support tidyeval's quasi-quotation in `.()` and perhaps in `:=`().
```
# desired query
DT[, .(sum_x = sum(x, na.rm=TRUE))]

# your proposed API
DT[, .(out_col_name = fun(in_col_name, fun_arg1=fun_arg1val)),
  env = list(in_col_name=in_col_name, fun=fun, fun_arg1=fun_arg1, fun_arg1val=fun_arg1val, out_col_name=out_col_name)]

# tidyeval would look much simpler and readable to me (variables would )
DT[, .(!!out_col_name := (!!fun)(!!in_col_name, !!fun_arg1 := !!fun_arg1val))]
```
The same would work in `by`/`keyby` and `on`.
For assignments in `j` one could either 
```
DT[, .`:=`(!!out_col_name := .((!!fun)(!!in_col_name, !!fun_arg1 := !!fun_arg1val))]
# which would substitute to:
# DT[, `:=`(sum_x = sum(x, na.rm=TRUE))]
```
or allow to use bare names in `.()` on the left hand side of `:=`.
Not sure about `i` though. In any case, I have not really thought this through and tidyeval might indeed not be mature enough (note, however, that I only suggested to use the syntax, not the package), I was really only wondering what speaks against this since it seems so much simpler from a user's (that is used to tidyeval) perspective.

**Comment by MichaelChirico** [comment]
@jan-glx thanks for your input.

understand there are different tastes but for me personally that version with all the `!!` does not look simple or intuitive at all.

one point worth mentioning about the `env` approach is that i think it aligns with a typical usage pattern for me, which is to keep these kind of metaparameters in a list for the whole analysis.

So for me, a more representative use case would be:

```
# at top of script
params = list(
  in_col_name=in_col_name, 
  fun=fun, 
  fun_arg1=fun_arg1, 
  fun_arg1val=fun_arg1val, 
  out_col_name=out_col_name,
  # ... other hyperparameters ...
)
# ... any other setup, data i/o, etc ...

DT[env = params, , .(out_col_name = fun(in_col_name, fun_arg1=fun_arg1val))]

# other queries using the same hyperparameters
DT[env = params, ...] 
```

**Comment by jangorecki** [comment]
I strongly agree with Michael about `!!`. This type of logical operations has its valid use cases, consider having integer variable that you want to `as.logical(var)`. How much more typing you need rather than `!!var`? I agree it is not a best practice but if you just play with data interactively, and not write a code or script, then `!!` is a good one.

**Comment by JohnMount** [comment]
I think `!!` is bad idea. It isn't in fact a no-op operation (see `!!c(0, 1, 2)`).  Also possibly the `rlang` team will eventually get a special glyph or parse change into the `R` language and then `data.table` will be stuck using an abandon convention. I think the `rlang` team did get a change into `R` on how `!!` itself is parsed to avoid something like `!(!())`.

**Comment by jangorecki** [comment]
@jan-glx a very fresh example of confusing use of `:=` from rlang: https://stackoverflow.com/a/63926023/2490497

**Comment by renkun-ken** [comment]
Most of my use case is programmatically determine the list to supply to `env=`, i.e.,

```r
dt[, X := Y + Z, env = list(X = paste0("x", i), Y = paste0("y", i), Z = paste0("z", i))
```

which looks much cleaner than

```r
X <- paste0("x", i)
Y <- paste0("y", i)
Z <- paste0("z", i)
dt[, !!X := !!Y + !!Z]
```

**Comment by jangorecki** [comment]
Another example of feature suggested in this PR can be seen on this SO question which asks for quite complex parametrizing of DT query: https://stackoverflow.com/questions/24833247/how-can-one-work-fully-generically-in-data-table-in-r-with-column-names-in-varia

**Comment by renkun-ken** [comment]
Very excited to see the PR is finally merged.
